src/lib.rs

```diff
@@ -198,7 +223,75 @@ impl Contract {
         notify::notify_mentions(desc.as_str(), id);
     }
 
+    #[payable]
+    pub fn add_proposal(&mut self, body: VersionedProposalBody, labels: HashSet<String>) {
+        near_sdk::log!("add_proposal");
+        let id = self.proposals.len();
+        let author_id = env::predecessor_account_id();
+        let editor_id = author_id.clone();
+        require!(
+            self.is_allowed_to_use_labels(
+                Some(editor_id.clone()),
+                labels.iter().cloned().collect()
+            ),
+            "Cannot use these labels"
+        );
+
+        for label in &labels {
+            let mut other_proposals = self.label_to_proposals.get(label).unwrap_or_default();
+            other_proposals.insert(id);
+            self.label_to_proposals.insert(label, &other_proposals);
+        }
+
+        let mut author_proposals = self.author_proposals.get(&author_id).unwrap_or_else(|| HashSet::new());
+        author_proposals.insert(id);
+        self.author_proposals.insert(&author_id, &author_proposals);
+
+        let proposal = Proposal {
+            id,
+            author_id: author_id.clone(),
+            block_height: 0u64, // TODO
+            snapshot: ProposalSnapshot { editor_id, timestamp: env::block_timestamp(), labels, body: body.clone() },
+            snapshot_history: vec![],
+        };
+
+        proposal::repost::publish_to_socialdb_feed(
+            Self::ext(env::current_account_id())
+            .with_static_gas(env::prepaid_gas().saturating_div(3))
+            .set_block_height_callback(proposal.clone()), proposal.clone());
+
+        let desc = get_proposal_description(body);
+
+        notify::notify_mentions(desc.as_str(), id);
+    }
+
+
+    #[private]
+    pub fn set_block_height_callback(&mut self, old_proposal: Proposal) -> u64 {
+        assert_eq!(env::promise_results_count(), 1, "ERR_TOO_MANY_RESULTS");
+        match env::promise_result(0) {
+            PromiseResult::Successful(val) => {
+                if let Ok(set_result) = near_sdk::serde_json::from_slice::<Value>(&val) {
+                    let block_height: u64 = set_result["block_height"].as_str().unwrap().parse::<u64>().unwrap();
+                    let proposal = Proposal {
+                        id: old_proposal.id,
+                        author_id: old_proposal.author_id.clone(),
+                        block_height: block_height,
+                        snapshot: old_proposal.snapshot,
+                        snapshot_history: old_proposal.snapshot_history,
+                    };
```

Comment:
You can define `old_proposal` as `mut` right in the function signature (since you expect the parameter to be passed "by value", it means that you will own it, and so it is totally fine to also mark it as internally mutable)

```rust
pub fn set_block_height_callback(&mut self, mut old_proposal: Proposal)
...

old_proposal.block_height = block_height;
self.proposals....
```